sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSqlCalciteTranslationUtils.java

```diff
@@ -17,174 +17,333 @@
+        return rexBuilder.makeLiteral(
+            value.getStringValue(), toCalciteType(type, false, rexBuilder), true);
+      case TYPE_BYTES:
+        return rexBuilder.makeBinaryLiteral(new ByteString(value.getBytesValue().toByteArray()));
+      case TYPE_NUMERIC:
+        // Cannot simply call makeExactLiteral() for ZetaSQL NUMERIC type because later it will be
+        // unparsed to the string representation of the BigDecimal itself (e.g. "SELECT NUMERIC '0'"
+        // will be unparsed to "SELECT 0E-9"), and Calcite does not allow customize unparsing of
+        // SqlNumericLiteral. So we create a wrapper function here such that we can later recognize
+        // it and customize its unparsing in BeamBigQuerySqlDialect.
+        return rexBuilder.makeCall(
+            SqlOperators.createZetaSqlFunction(
+                BeamBigQuerySqlDialect.NUMERIC_LITERAL_FUNCTION,
+                toCalciteType(type, false, rexBuilder).getSqlTypeName()),
+            rexBuilder.makeExactLiteral(
+                value.getNumericValue(), toCalciteType(type, false, rexBuilder)));
+      case TYPE_DATE:
+        return rexBuilder.makeDateLiteral(dateValueToDateString(value));
+      case TYPE_TIME:
+        return rexBuilder.makeTimeLiteral(
+            timeValueToTimeString(value),
+            rexBuilder.getTypeFactory().getTypeSystem().getMaxPrecision(SqlTypeName.TIME));
+      case TYPE_DATETIME:
+        return rexBuilder.makeTimestampWithLocalTimeZoneLiteral(
+            datetimeValueToTimestampString(value),
+            rexBuilder
+                .getTypeFactory()
+                .getTypeSystem()
+                .getMaxPrecision(SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE));
+      case TYPE_TIMESTAMP:
+        return rexBuilder.makeTimestampLiteral(
+            timestampValueToTimestampString(value),
+            rexBuilder.getTypeFactory().getTypeSystem().getMaxPrecision(SqlTypeName.TIMESTAMP));
+      case TYPE_ARRAY:
+        return arrayValueToRexNode(value, rexBuilder);
+      case TYPE_STRUCT:
+        return structValueToRexNode(value, rexBuilder);
+      case TYPE_ENUM: // internal only, used for DateTimestampPart
+        return enumValueToRexNode(value, rexBuilder);
+      default:
+        throw new UnsupportedOperationException("Unknown ZetaSQL type: " + type.getKind().name());
+    }
+  }
+
+  private static RexNode arrayValueToRexNode(Value value, RexBuilder rexBuilder) {
+    return rexBuilder.makeCall(
+        toCalciteArrayType(value.getType().asArray().getElementType(), false, rexBuilder),
+        SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR,
+        value.getElementList().stream()
+            .map(v -> toRexNode(v, rexBuilder))
+            .collect(Collectors.toList()));
+  }
+
+  private static RexNode structValueToRexNode(Value value, RexBuilder rexBuilder) {
+    return rexBuilder.makeCall(
+        toCalciteStructType(value.getType().asStruct(), false, rexBuilder),
```

Comment:
This LOC specified the return STRUCT type explicitly, so Calcite does not need to infer. Adding this line introduced a bug, because later when the schema is translated to Beam Schema, duplicate field names will be caught. https://github.com/apache/beam/pull/12550 fixes this bug.

@apilloud